src/material/core/datetime/date-selection.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DateAdapter} from '@angular/material/core';
+import {Subject} from 'rxjs';
+
+export abstract class MatDateSelection<D> {
+  valueChanges = new Subject<void>();
+
+  constructor(protected readonly adapter: DateAdapter<D>) {}
+
+  dispose() {
+    this.valueChanges.complete();
+  }
+
+  abstract add(date: D | null): void;
+  abstract clone(): MatDateSelection<D>;
+  abstract getFirstSelectedDate(): D|null;
+  abstract getLastSelectedDate(): D|null;
+  abstract isComplete(): boolean;
+  abstract isSame(other: MatDateSelection<D>): boolean;
+  abstract isValid(): boolean;
+}
+
+export interface DateRange<D> {
+  start: D | null;
+  end: D | null;
+}
+
+/**
+ * Concrete implementation of a MatDateSelection that holds a single date.
+ */
+export class MatSingleDateSelection<D> extends MatDateSelection<D> {
+  private date: D | null = null;
+
+  constructor(adapter: DateAdapter<D>, date?: D | null) {
+    super(adapter);
+    if (date) {
+      this.date = date;
+    }
+  }
+
+  add(date: D | null) {
+    this.date = date;
+    this.valueChanges.next();
+  }
+
+  compareDate(other: MatSingleDateSelection<D>) {
+    const date = this.asDate();
+    const otherDate = other.asDate();
+    if (date && otherDate) {
+      return this.adapter.compareDate(date, otherDate);
+    } 
+    throw Error
+
+  }
+
+  clone(): MatDateSelection<D> {
+    return new MatSingleDateSelection<D>(this.adapter, this.date);
+  }
+
+  getFirstSelectedDate() { return this.date; }
+
+  getLastSelectedDate() { return this.date; }
+
+  isComplete() { return !!this.date; }
+
+  isSame(other: MatDateSelection<D>): boolean {
+    return other instanceof MatSingleDateSelection &&
+        this.adapter.sameDate(other.getFirstSelectedDate(), this.getFirstSelectedDate());
+  }
+
+  isValid(): boolean {
+    return !!(this.date && this.adapter.isDateInstance(this.date) && this.adapter.isValid(this.date));
+  }
+
+  asDate(): D | null {
```

Comment:
if we parameterize on `D` and `S` this could be a base class method called `getSelection` - again, worth keeping in mind in case we go that way